### DISCUSSION TITLE
Trouble understanding double-length numbers

### DISCUSSION THREAD

**Discussion by ChristopherDrum** [jkotlinski/durexforth#372] [Q&A] [answered]
Absolute Forth beginner here, mostly working my way through Starting FORTH (as recommended by the README). At any rate, various FORTH manuals (Starting FORTH and A Beginner's Guide to FORTH) talk about double-length numbers as though they are quite simple to work with.

For example, Starting FORTH says, "Some Forth implementations (including SwiftForth) will convert any number that contains the following characters as a double number: + , - . / : " and Beginner's Guide says "A double-length integer is entered from the keyboard by including a decimal point in it."

So, at the least both texts seem to agree that adding a `.` to an input number will be converted into a double-length number and placed on the stack as such. However, this does not seem to be the case for DurexForth. 

What is the equivalent in DurexForth?

**Comment by jkotlinski** [comment]
Hi!
Durexforth does not support parsing double numbers by adding decimal point.
Instead of typing ”1234.” , you would have to type ”1234 0” (or is it other
way around? Don’t remember right now.)

**Reply by ChristopherDrum** [reply]
So, yes, I suppose to state the question more simply, "How do I put a double-length number on the stack?"

**Reply by ChristopherDrum** [reply]
Hmm... according to the DurexForth manual, we should look to "Starting FORTH" to understand number formatting. It was in that section of Starting that I came upon this double-length issue. For example, in the case of a phone number that book simply enters 345-6789 and the hyphen forces Forth to parse it into a double-length number.

I see there is a `>number` in the compat module, which seems like it would work to make a phone number string into an unsigned double. Maybe that is how to handle these situations in Durex, but I'm not yet clear how to work with strings so I'm getting ahead of myself.

**Reply by ChristopherDrum** [reply]
OK, I got a little success with your description. It is "the other way around". For the 32-bit double-length number, put the least significant bits first as their own value, then the most significant bits.
So `123456` can be put onto the stack as hex `e240 01` which can be reconstituted into decimal `123456` using `<# #s #> type`
Still a lot to learn, especially how to use this knowledge to handle double-length input as a user would want to enter it, but at least the formatting issue is working as expected.

**Comment by polluks** [comment]
Indeed, the Double-Number word set is optional
http://lars.nocrew.org/forth2012/double.html

**Reply by ChristopherDrum** [reply]
Ah, interesting. Thanks for pointing me toward that information.

**Reply by JoannaK** [reply]
Forth2012 standard is also downloadable as a PDF for local read, 260 pages. 
http://www.forth200x.org/documents/forth-2012.pdf

**Reply by JoannaK** [reply]
I think there should be a note somewhere, that Starting Forth while really good introduction, is an ancient book and does not fully follow current Forth standards. It was written like 40 years ago, in era of Fig-forth and forth-79. Since then there has been at least -83, Ansi, -94 and Forth2012 standards afterwards, it's feasible that some changes have been made.

Also to be noted that Forth.inc had altered to make online version fit their own products. Especially chapter 7 (numbers) has been partially written to look like 32-bit forth and so a lot more confusing for 16-bit forth users.

**Comment by Whammo** [comment, marked as answer]
#445 has some good examples of using longs.
